demos/blog/setup.py

```diff
@@ -1,6 +1,20 @@
 from setuptools import setup
```

Comment:
It's not a good idea to have `setup.py` in an app. It's meant for libraries.

Reply:
Better count everything as a library. For instance, your tests will use your app as a library to test it.

Reply:
Don't confuse distribution package with library.

Reply:
No differences at all, actually.

Reply:
I we follow your claims we need a separate `setup.py` in `tests/` folder

Reply:
Why? Why not to make blog app as a regular project with setup.py / pyproject.toml? Same rules runs for all types, no differences leads to simplify their support and management. You still need some of those to describe app dependencies, build routines, etc.

Reply:
It depends on the purpose: if you want to distribute an app which users will call, while still being able to configure their own db, static files etc in a flexible manner - setuptools fits there. If it's a typical `git clone` and `pip install -e .` - it's odd to combine those unrelated approaches.

As for `pyproject.toml`, it's not the same. It's more of a framework for building standardized packaging, which only reserves build deps, while omitting any clear recommendations on how other tools should maintain their own sections.

P.S. Since approximately pip v10 it supports `pyproject.toml`s build deps, which it installs before invoking `setup.py`.

Reply:
It doesn't matters generates it or not. The matter is that project provides some file that contains project metadata and allows to use it with some Python package management tool to built it, develop it, manage it and so on.

I don't see any reasons to use development mode in production, for sure. But I do want it for development. How do you develop any of your app? I guess, you do `pip install -e .` or kinda of that and follow your regular workflow. 

- Why do you want to remove ability to develop an app from it source code? 
- Why don't you want to share common development, packaging and distributing experience with the all the rest ones? 
- Why to make things complicated so there could be libraries and apps?

Library could easily became an app - just give it some entry point (cli, web ui, etc). App could easily be a library - to import some stuff to extend it.

P.S. I feel we're going to make long discussion about that topic, so how about to continue it somewhere in different place? It doesn't really about this PR, but our views on the things.

Reply:
It always depends on the use case. But here I wanted to point out that the author mixes up a few tools, which work with different levels of env ecosystem.

Oh, and your assumptions about my intentions are completely wrong. I don't want to limit someone's ability to do development. I want people to step onto the same rake I did a while back.
I'm exploring the packaging ecosystem for a few years now and atm there's no *good* standard for doing things.

There's setuptools-based approach to distributing software. It takes care of putting some files into env (site-packages), but that's it. It also triggers dependencies installation. **It mutates current env.**
There's pip, which essentially calls `setup.py install` or `setup.py develop` (`-e` option), or unpacks wheels (of course, it's a high-level overview and generalization, but details don't matter at this point). It also mutates current env.

Now, you write an application, so you need to pin deps of the whole env and you need a separate (virtual)env, where you'll only have yourself influencing it and nobody else touching what's inside.

setuptools+pip doesn't solve this in any way. There's a practice of non-locking deps in setup.py, because it's intended to be used for libs, which will be used by applications: you don't want libs to conflict with each other because of too tight restrictions.

The solution is to have some description for the env, ideally, you should have ability to do encapsulated/isolated installation of that.
requirements.txt partially solves that, pipenv and tox solve it even better. there's also different kinds of less popular packagers, which allow you to bundle everything as one archive.

But this usually assumes that you'll deliver source to the file system somehow, create a venv, install requirements and will go from there. See? No need to "install" the project itself with pip. Some will argue that it's unpure to hijack the path (pythonpath), but it depends. Some will use `pip install -e .` to make stuff importable instead of doing that.

I think, there's no need to actually use setuptools when you don't need to distribute the thing. There's a use-case for apps, where you need it - that is when you want to ship some script accessible from the venv by its name. It's a completely different model and currently I don't see any need to actually use it.

*Remark:* a lot of people are actually trying to get rid of setuptools because of broken design decisions it introduced a decade ago, which are impossible to fix because of backwards compatibility. It's actually harmful in many ways (any broken dist may break yours, there's heavy import time operations etc.)

P.S. I'd like to empathize that there's no single standard for doing things right wrt to packaging in Python, but you might want to keep an eye on the Python Packaging Authority (PyPA), who are trying to make things better: https://pypa.io.

Interesting refs:
* https://caremad.io/posts/2013/07/setup-vs-requirement/
* https://www.pypa.io/en/latest/history/
* https://packaging.python.org/

P.S. Feel free to open an issue for discussion if you're interested.

Reply:
@kxepal you might be interested in https://snarky.ca/clarifying-pep-518/ also as it became more relevant for modern envs :)

Reply:
Fresh one, as of today: https://twitter.com/ncoghlan_dev/status/1028963099909353472?s=19

Reply:
@webknjaz 
Thanks a lot for such detailed post and links! I need some time to think about all of this.